### PR TITLE
adds Module Code

### PR DESCRIPTION
**Pull request by ksatirli**
Initial version of this module.

suggested merge message:

```
* adds upstream `aws-ia` configurations
  * updates `CODEOWNERS`

* adds basic example and documentation

* adds module code

* adds documentation for base module
```

### CODE REVIEW

**Comment by ksatirli** [comment]
@drewmullen could I get an initial review of this?

Specifically looking for missing best-practices for S3, CF

File: README.md
Comment (drewmullen) [inline README.md:1]: Is it worth including a diagram to describe the deployment?

File: README.md
Comment (drewmullen) [inline README.md]: `operatehappy`

File: USAGE.md
Comment (drewmullen) [inline USAGE.md]: what is this file?

File: examples/basic/.envrc.sample
Comment (drewmullen) [inline examples/basic/.envrc.sample]: should avoid having users use static keys

File: examples/basic/providers.tf
Comment (drewmullen) [inline examples/basic/providers.tf]:
```
$ gh api repos/aws-ia/terraform-awscc-dropshare | jq .id
459237564
```

File: examples/basic/providers.tf
Comment (drewmullen) [inline examples/basic/providers.tf]: placeholder to talk about module name again

File: networking.tf
Comment (drewmullen) [inline networking.tf]: did we get any direction from @netdevautomate on this?

**Review by drewmullen — DISMISSED** [review]
looks fantastic. will get back to you on github actions and we need to talk about the name `-module`

File: README.md
Comment (ksatirli) [reply README.md:1]: With mermaid, yes?

File: README.md
Comment (ksatirli) [reply README.md]: Great catch - fixed!

File: USAGE.md
Comment (ksatirli) [reply USAGE.md]: hmm - this must have been created by `terraform-docs`' GHA. removing it!

File: examples/basic/providers.tf
Comment (ksatirli) [reply examples/basic/providers.tf]: fixed!

File: networking.tf
Comment (ksatirli) [reply networking.tf]: Not yet!

File: examples/basic/.envrc.sample
Comment (ksatirli) [reply examples/basic/.envrc.sample]: I can remove the .sample file, indeed!

File: README.md
Comment (ksatirli) [reply README.md:1]: added a diagram

File: networking.tf
Comment (drewmullen) [reply networking.tf]: can you add him to this fork?

**Comment by ksatirli** [comment]
@drewmullen could I get another review? 

Also: @NetDevAutomate, if you have thoughts on CF policies and IAM policies, I'll take them!

File: networking.tf
Comment (ksatirli) [reply networking.tf]: done! @NetDevAutomate any thoughts?

**Comment by ksatirli** [comment]
@drewmullen specifically would love some guidance on a better way to express the IAM config to avoid [this error](https://github.com/ksatirli/terraform-awscc-dropshare/security/code-scanning/1?query=ref%3Arefs%2Fpull%2F1%2Fmerge+ref%3Arefs%2Fpull%2F1%2Fhead+ref%3Arefs%2Fheads%2Fadds-module-code)

File: .pre-commit-config.yaml
Comment (drewmullen) [inline .pre-commit-config.yaml]: we discovered recently that you also need to prepend `__GIT_WORKING_DIR__` to tflint as well

File: .snyk
Comment (drewmullen) [inline .snyk:8]: Ive never looked at this tool but the fact that you can log the reason to ignore is fantastic. Its actually an idea i passed over to `tfsec` also: https://github.com/aquasecurity/tfsec/issues/1416

File: CODEOWNERS
Comment (drewmullen) [inline CODEOWNERS]: In I&A codeowners will have to be me and tony v

File: README.md
Comment (drewmullen) [inline README.md:39]: This should be slightly expanded to describe what will the end user get. ie. "you'll be able to connect your dropshare app to a fully functional, best practice, aws backend"... something like that

File: README.md
Comment (drewmullen) [inline README.md:58]: perhaps an inline comment with a link to the dropshare docs for how to generate this secret key?

File: README.md
Comment (drewmullen) [inline README.md:59]: <3 for including this

File: README.md
Comment (drewmullen) [inline README.md:88]: should this default be there? `Z2FDTNDATAQYW2`

File: README.md
Comment (drewmullen) [inline README.md:127]: We will allow the `.workflow` dir in I&A but have disabled github actions. Our CI is being refactored to make use of a Repos pre-commit. perhaps adjust this section to mention the pre-commit and not github actions

File: variables.tf
Comment (drewmullen) [inline variables.tf:16]: what is this default for?

File: examples/basic/providers.tf
Comment (drewmullen) [inline examples/basic/providers.tf:36]: tags should be a variable

if you want to easily use the same tags for `awscc_` resources pass the new `var.tags` to the [label module](https://github.com/aws-ia/terraform-aws-label)

```terraform
module "tags" {
  source  = "aws-ia/label/aws"
  version = "0.0.4"

  tags = var.tags
}
```

then you can reference in awscc resources like (example concat'd with a static tag value):

```terraform
  tags = concat(
    [{"key" = "Name", "value" = var.name}],
      module.tags.tags
  )
```

You can always refer to these tags for the regular aws provider also using
`module.tags.tags_aws` but the use of `default_tags` probably negates any use for that in this module

File: examples/basic/providers.tf
Comment (drewmullen) [inline examples/basic/providers.tf]: update this name

**Review by drewmullen — CHANGES_REQUESTED** [review]
Looking really good!

One thing you should include in the initial release is at least 1 test to build your examples/basic root. With awscc still being alpha having a quick way to test each module is paramount

File: .snyk
Comment (ksatirli) [reply .snyk:8]: Agreed! Makes auditing much much easier.

File: CODEOWNERS
Comment (ksatirli) [reply CODEOWNERS]: We can fix this after the move *nods*

File: README.md
Comment (ksatirli) [reply README.md:58]: The secret key is generated through this module and then encrypted with a Keybase key and made available through an output.

File: README.md
Comment (ksatirli) [reply README.md:88]: In the docs... probably not. 

Is there a default/ example Zone ID I could use? (Similarly to the EXAMPLE IAM key)

File: variables.tf
Comment (ksatirli) [reply variables.tf:16]: Omitting the name will result in a random bucket name being generated.

**Comment by drewmullen** [comment]
> @drewmullen specifically would love some guidance on a better way to express the IAM config to avoid [this error](https://github.com/ksatirli/terraform-awscc-dropshare/security/code-scanning/1?query=ref%3Arefs%2Fpull%2F1%2Fmerge+ref%3Arefs%2Fpull%2F1%2Fhead+ref%3Arefs%2Fheads%2Fadds-module-code)[](https://github.com/ksatirli)

isnt the point of the module to allow dropshare access to a bucket? I'm not sure we care much about that tbh

File: variables.tf
Comment (drewmullen) [reply variables.tf:16]: cool - can you update the description to include that behavior?

File: README.md
Comment (drewmullen) [reply README.md:88]: perhaps just no default and have an example in the description?

File: .pre-commit-config.yaml
Comment (ksatirli) [inline .pre-commit-config.yaml]:
```suggestion
          - "--args=--config=__GIT_WORKING_DIR__/.tflint.hcl"
```

File: README.md
Comment (ksatirli) [reply README.md:88]: My mistake - `Z2FDTNDATAQYW2` is the magical ID that is used for _all_ CloudFront links, so this _is_ intentional.

File: README.md
Comment (ksatirli) [reply README.md:127]: Works for me. I'd vote to keep the files as they are solid implementations of these checks (and easier for members of the community to see _what_ is happening and how it is checked)

File: variables.tf
Comment (ksatirli) [reply variables.tf:16]: Done

File: examples/basic/providers.tf
Comment (ksatirli) [reply examples/basic/providers.tf:36]: I believe `default_tags` isn't eligible for variables, as it is in the `providers` block.

On a per-resource basis, it would be possible to use this module.

File: examples/basic/providers.tf
Comment (ksatirli) [reply examples/basic/providers.tf]: Done across the repository

**Comment by ksatirli** [comment]
> > @drewmullen specifically would love some guidance on a better way to express the IAM config to avoid [this error](https://github.com/ksatirli/terraform-awscc-dropshare/security/code-scanning/1?query=ref%3Arefs%2Fpull%2F1%2Fmerge+ref%3Arefs%2Fpull%2F1%2Fhead+ref%3Arefs%2Fheads%2Fadds-module-code)
> 
> isnt the point of the module to allow dropshare access to a bucket? I'm not sure we care much about that tbh

It is, but I believe the IAM policy can be written in a stronger way. Just not entirely sure _how_ :D

File: .pre-commit-config.yaml
Comment (drewmullen) [inline .pre-commit-config.yaml:37]: this is redundant, `-recursive` runs by default for that hook

File: .pre-commit-config.yaml
Comment (drewmullen) [inline .pre-commit-config.yaml:42]: by setting this explicitly, you _only_ validate the examples. Is that the intention? curious of your thought patterns here

File: .pre-commit-config.yaml
Comment (drewmullen) [inline .pre-commit-config.yaml:55]: the effect of declaring this twice is that the pre-commit looks for a `.terraform-docs.yml` file in `examples/basic`. That is a cool idea but:
1. I dont see one so im wondering how the setup is working for you. i get an error when i try this
2. If we're going to declare a specific .terraform-docs.yaml in each examples then we should suppress #L45-L48 from overwriting the file

**Comment by drewmullen** [comment]
> It is, but I believe the IAM policy can be written in a stronger way. Just not entirely sure how :D


i have no ideas, or concern for increasing least priv for that policy. IMO the point of the bucket is for dropshare to have access so it seems like a good policy to me

File: .pre-commit-config.yaml
Comment (ksatirli) [reply .pre-commit-config.yaml:42]: The reason I validate from `examples` is that in there, variables for the module are set and `terraform init` has created a lock file.

If I validate on the top level of the module, I get errors like:

> Error: Provider configuration not present

as provider-specific config needs to be passed to the module. Since that has to be done _outside_ the module, I think it is correct to test from `examples/basic`, vs from top-level

File: .pre-commit-config.yaml
Comment (ksatirli) [reply .pre-commit-config.yaml:55]: This is a strange one, indeed! I've generated the file a few times and it (seemed to?) work.

I'll investigate.

Comment (ksatirli) [reply .pre-commit-config.yaml:55]: I'm unable to reproduce this as not being generated. I'll close this for now but am happy to re-open if you have more pointers @drewmullen

File: README.md
Comment (drewmullen) [inline README.md:35]: we gotta figure out what to do with this:

https://www.youtube.com/watch?v=i2l4Wx9YZ1E

**Review by drewmullen — CHANGES_REQUESTED** [review]
only thing left IMO is what to do with mermaid diagram

File: README.md
Comment (ksatirli) [reply README.md:35]: I recommend keeping this, as it provides a simple overview in textual form

**Pull request merged**
